modules/apps/oauth2-provider/oauth2-provider-scope-impl/src/main/java/com/liferay/oauth2/provider/scope/internal/graphql/OAuth2GraphQLRequestContextValidator.java

```diff
@@ -0,0 +1,199 @@
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.oauth2.provider.scope.internal.graphql;
+
+import com.liferay.oauth2.provider.scope.ScopeChecker;
+import com.liferay.oauth2.provider.scope.liferay.OAuth2ProviderScopeLiferayAccessControlContext;
+import com.liferay.oauth2.provider.scope.liferay.ScopeContext;
+import com.liferay.portal.kernel.security.access.control.AccessControlUtil;
+import com.liferay.portal.kernel.security.access.control.AccessControlled;
+import com.liferay.portal.kernel.security.auth.AccessControlContext;
+import com.liferay.portal.kernel.security.auth.verifier.AuthVerifierResult;
+import com.liferay.portal.kernel.security.service.access.policy.ServiceAccessPolicy;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.security.access.control.AccessControlAdvisor;
+import com.liferay.portal.security.access.control.AccessControlAdvisorImpl;
+import com.liferay.portal.vulcan.graphql.validation.GraphQLRequestContext;
+import com.liferay.portal.vulcan.graphql.validation.GraphQLRequestContextValidator;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import javax.ws.rs.ForbiddenException;
+import javax.ws.rs.core.Application;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+/**
+ * @author Carlos Correa
+ */
+@Component(service = GraphQLRequestContextValidator.class)
+public class OAuth2GraphQLRequestContextValidator
+	implements GraphQLRequestContextValidator {
+
+	@Override
+	public void validate(GraphQLRequestContext graphQLRequestContext)
+		throws Exception {
+
+		_scopeContext.setApplicationName(
+			graphQLRequestContext.getApplicationName());
+		_scopeContext.setBundle(
+			FrameworkUtil.getBundle(graphQLRequestContext.getResourceClass()));
+		_scopeContext.setCompanyId(graphQLRequestContext.getCompanyId());
+
+		if (OAuth2ProviderScopeLiferayAccessControlContext.
+				isOAuth2AuthVerified()) {
+
+			_enableSAP(graphQLRequestContext.getApplicationName());
+
+			if (!_scopeChecker.checkScope(
+					graphQLRequestContext.getHttpMethod())) {
```

Comment:
This assumes that the JAX-RS application is using HTTP verbs for scopes (`HttpMethodFeature`). This is the default for JAX-RS applications, and also what REST Builder generated ones use. However we should support other ones also. Most notably annotation based scopes `AnnotationFeature` .